packages/completer/src/handler.ts

```diff
@@ -477,6 +486,85 @@ export namespace CompletionHandler {
     connector: IDataConnector<IReply, void, IRequest>;
   }
 
+  /**
+   * Wrapper object for ICompletionItem.
+   * Implementers of this interface should be responsible for
+   * deduping and sorting the items in the list.
+   */
+  export interface ICompletionItems {
+    /**
+     * Collection of completion items.
+     */
+    items: Array<ICompletionItem>;
+  }
+
+  /**
+   * Completion item object based off of LSP CompletionItem
+   */
+  export interface ICompletionItem extends PartialJSONObject {
+    /**
+     * User facing completion.
+     * If insertText is not set, this will be inserted.
+     */
+    label: string;
+
+    /**
+     * Completion to be inserted.
+     */
+    insertText?: string;
+
+    /**
+     * Range to be replaced by this completion.
+     */
+    range?: IRange;
+
+    /**
+     * Type of this completion item.
+     */
+    type?: string;
+
+    /**
+     * Image url for icon to be rendered with completion type.
+     */
+    icon?: string;
```

Comment:
I have some questions about this:

- Where is this coming from, and where does it actually get used?
- Does it have to be implemented as an image url string?

Across the jlab codebase we use a consistent naming scheme for icon-related data in interfaces:

- `icon`: an instance of `LabIcon`, or of something that can be resolved to a `LabIcon`
- `iconClass`: the name of a class that has an associated CSS rule that sets some icon as the `background-image`

We don't take raw image urls as icons anywhere else, so ideally the above would get changed to eg `iconClass`. At the very least, this field needs to be called something other than `icon`

Reply:
> Where is this coming from, and where does it actually get used?

The purpose of the field is to display an icon given by completion providers next to items in the completer list. There's more information on this in #8140.

> Does it have to be implemented as an image url string?

SVG data-uris seemed like the simplest approach (easily serialized, doesn't require separate fetching/dereferencing).

However, given that extensions can provide styling and thereby add icons via custom `iconClass`es, that approach seems good as well. Providing icons that are generated on-the-fly would be much harder or maybe impossible, but I don't think that's a big use-case, and passing class names around instead of SVG data will make completion responses smaller. I will switch to using that method instead.

Reply:
@saulshanabrook and I just had a discussion about this. For now, the interface for icons here should be just:

```typescript
  icon?: LabIcon
```

You would then implement the creation of the icon DOM node with:

```typescript
if (icon) {
  const iconNode = icon.element({className: 'jp-Completer-type'});
  li.appendChild(iconNode);
}
```

Reply:
There's an argument to be made that there should be both an `icon` and an `iconClass` field here, but we can wait to hash that out until #8237 is resolved

Reply:
This approach makes a lot of sense to me, and I have made the changes here and in #8080